From f0d764ada1a9a39a64ea9d95c30c6bee49789fd8 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:49:07 -0400
Subject: [fs] nfsd: racy access to ->d_name in nsfd4_encode_path()

Message-id: <1403560314-32525-6-git-send-email-bfields@redhat.com>
Patchwork-id: 84447
O-Subject: [Patch RHEL7 005/172] nfsd: racy access to ->d_name in nsfd4_encode_path()
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Al Viro <viro@zeniv.linux.org.uk>

Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Upstream: 301f0268b63d1b07268e46f5901fc51d6cac20eb
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index c6c727a..ecc735e 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -1818,10 +1818,7 @@ static __be32 nfsd4_encode_fs_location4(struct nfsd4_fs_location *location,
 static __be32 nfsd4_encode_path(const struct path *root,
   const struct path *path, __be32 **pp, int *buflen)
 {
- struct path cur = {
-  .mnt = path->mnt,
-  .dentry = path->dentry,
- };
+ struct path cur = *path;
  __be32 *p = *pp;
  struct dentry **components = NULL;
  unsigned int ncomponents = 0;
@@ -1861,14 +1858,19 @@ static __be32 nfsd4_encode_path(const struct path *root,
 
  while (ncomponents) {
   struct dentry *dentry = components[ncomponents - 1];
-  unsigned int len = dentry->d_name.len;
+  unsigned int len;
 
+  spin_lock(&dentry->d_lock);
+  len = dentry->d_name.len;
   *buflen -= 4 + (XDR_QUADLEN(len) << 2);
-  if (*buflen < 0)
+  if (*buflen < 0) {
+   spin_unlock(&dentry->d_lock);
    goto out_free;
+  }
   WRITE32(len);
   WRITEMEM(dentry->d_name.name, len);
   dprintk("/%s", dentry->d_name.name);
+  spin_unlock(&dentry->d_lock);
   dput(dentry);
   ncomponents--;
  }
-- 
1.7.1